### PR TITLE
tcp/unix input: Stop accepting connections after socket is closed

### PR DESCRIPTION
**Pull request by andrewkroh**
## What does this PR do?

Stop accepting connections after tcp/unix socket is closed. This will suppress debug messages for net.ErrClosed.

    2021-12-22T13:49:04.151Z	INFO	[tcp]	streaming/listener.go:172	StoppingTCPserver	{"address": "0.0.0.0:7000"}
    2021-12-22T13:49:04.151Z	DEBUG	[tcp]	streaming/listener.go:129	Can not accept the connection	{"address": "0.0.0.0:7000", "error": "accept tcp [::]:7000: use of closed network connection"}

## Why is it important?

This makes the shutdown process a little bit cleaning from a logging perspective.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Logs

Sample log output of a starting, accepting a connection, disconnecting, and shutting down.

```
{
  "log.level": "info",
  "@timestamp": "2022-01-05T15:56:53.293-0500",
  "log.logger": "syslog",
  "log.origin": {
    "file.name": "syslog/input.go",
    "file.line": 147
  },
  "message": "Starting Syslog input",
  "service.name": "filebeat",
  "protocol": "tcp",
  "ecs.version": "1.6.0"
}
{
  "log.level": "debug",
  "@timestamp": "2022-01-05T15:56:53.293-0500",
  "log.logger": "tcp",
  "log.origin": {
    "file.name": "streaming/listener.go",
    "file.line": 141
  },
  "message": "Start accepting connections",
  "service.name": "filebeat",
  "address": "127.0.0.1:7000",
  "ecs.version": "1.6.0"
}
{
  "log.level": "debug",
  "@timestamp": "2022-01-05T15:56:58.775-0500",
  "log.logger": "tcp",
  "log.origin": {
    "file.name": "streaming/listener.go",
    "file.line": 185
  },
  "message": "New client connection.",
  "service.name": "filebeat",
  "address": "127.0.0.1:7000",
  "remote_address": "127.0.0.1:62085",
  "active_clients": 1,
  "ecs.version": "1.6.0"
}
{
  "log.level": "debug",
  "@timestamp": "2022-01-05T15:56:58.776-0500",
  "log.logger": "tcp",
  "log.origin": {
    "file.name": "streaming/listener.go",
    "file.line": 188
  },
  "message": "Client disconnected.",
  "service.name": "filebeat",
  "address": "127.0.0.1:7000",
  "remote_address": "127.0.0.1:62085",
  "active_clients": 0,
  "ecs.version": "1.6.0"
}
{
  "log.level": "debug",
  "@timestamp": "2022-01-05T15:57:02.287-0500",
  "log.logger": "tcp",
  "log.origin": {
    "file.name": "streaming/listener.go",
    "file.line": 200
  },
  "message": "Stopping socket listener. Waiting for all connections to close.",
  "service.name": "filebeat",
  "address": "127.0.0.1:7000",
  "ecs.version": "1.6.0"
}
{
  "log.level": "debug",
  "@timestamp": "2022-01-05T15:57:02.287-0500",
  "log.logger": "tcp",
  "log.origin": {
    "file.name": "streaming/listener.go",
    "file.line": 151
  },
  "message": "Stopped accepting connections",
  "service.name": "filebeat",
  "address": "127.0.0.1:7000",
  "ecs.version": "1.6.0"
}
{
  "log.level": "info",
  "@timestamp": "2022-01-05T15:57:02.287-0500",
  "log.logger": "tcp",
  "log.origin": {
    "file.name": "streaming/listener.go",
    "file.line": 203
  },
  "message": "Socket listener stopped",
  "service.name": "filebeat",
  "address": "127.0.0.1:7000",
  "ecs.version": "1.6.0"
}
```

### CODE REVIEW

File: filebeat/inputsource/common/streaming/listener.go [hunk 1]
```diff
@@ -62,6 +64,23 @@ var (
...
+	ft, ok := framingTypes[value]
+	if !ok {
+		availableTypes := make([]string, len(framingTypes))
+		i := 0
+		for t := range framingTypes {
+			availableTypes[i] = t
+			i++
+		}
+		return fmt.Errorf("invalid framing type '%s', supported types: %v", value, availableTypes)
+
```
Comment (efd6) [inline filebeat/inputsource/common/streaming/listener.go:78]: `framingTypes` is never mutated AFAICS, so maybe the format string here could be constructed in an `init`?

Also I recall seeing recently that the indexed slice from map construction used here is no longer more efficient than the append-into-correctly-capped construction. If an `init` is used even that care is probably not worth the readability cost.

File: filebeat/inputsource/common/streaming/listener.go [hunk 2]
```diff
@@ -117,73 +138,72 @@ func (l *Listener) initListen(ctx context.Context) error {
 }
 
 func (l *Listener) run() {
-	l.log.Info("Started listening for " + l.family.String() + " connection")
+	l.log.Debug("Start accepting connections")
+	defer l.log.Debug("Stopped accepting connections")
+
+	defer l.Listener.Close()
```
Comment (efd6) [inline filebeat/inputsource/common/streaming/listener.go:144]: Bundle these into a single `defer func() {...`?

Comment (andrewkroh) [reply filebeat/inputsource/common/streaming/listener.go]: Done

File: filebeat/inputsource/common/streaming/listener.go (hunk 1 above)
Comment (andrewkroh) [reply filebeat/inputsource/common/streaming/listener.go]: Done (This code was only moved up from the bottom to be next to the type declaration.)

**Comment by andrewkroh** [comment]
Looks like I broke some system tests that look for particular log messages.

File: filebeat/inputsource/common/streaming/listener.go [hunk 3]
```diff
@@ -23,8 +23,11 @@ import (
 	"context"
 	"fmt"
 	"net"
+	"strings"
 	"sync"
 
+	"github.com/pkg/errors"
```
Comment (efd6) [inline filebeat/inputsource/common/streaming/listener.go:29]: Tangentially related, github.com/pkg/errors has been archived, does work that it looks like we never use and is arguably harder to reason about than the equivalent offering in stdlib. A little while ago I looked into what would be required to migrate to pure stdlib error handling with the same functionality and 99.9% can be done mechanically over all of beats without trying hard, but the change is on the order of 5,000 lines across 500 files.

File: filebeat/inputsource/common/streaming/listener.go [hunk 4]
```diff
@@ -60,8 +63,30 @@ var (
 		"delimiter": FramingDelimiter,
 		"rfc6587":   FramingRFC6587,
 	}
+
+	framingTypeNames []string
```
Comment (efd6) [inline filebeat/inputsource/common/streaming/listener.go:67]: `availableFramingTypesErrFormat string`

File: filebeat/inputsource/common/streaming/listener.go [hunk 5]
```diff
@@ -60,8 +63,30 @@ var (
...
 )
 
+func init() {
+	framingTypeNames := make([]string, 0, len(framingTypes))
+	for t := range framingTypes {
+		framingTypeNames = append(framingTypeNames, t)
+	}
+}
```
Comment (efd6) [inline filebeat/inputsource/common/streaming/listener.go:75]: `availableFramingTypesErrFormat = fmt.Sprintf("invalid framing type %%q, the supported types are [%v]", strings.Join(framingTypeNames, ", "))`

File: filebeat/inputsource/common/streaming/listener.go [hunk 6]
```diff
@@ -60,8 +63,30 @@ var (
...
+
+// Unpack unpacks the FramingType string value.
+func (f *FramingType) Unpack(value string) error {
+	value = strings.ToLower(value)
+
+	ft, ok := framingTypes[value]
+	if !ok {
+		return fmt.Errorf("invalid framing type %q, the supported types are [%v]", value, strings.Join(framingTypeNames, ", "))
```
Comment (efd6) [inline filebeat/inputsource/common/streaming/listener.go:83]: `return fmt.Errorf(availableFramingTypesErrFormat, value)`

File: filebeat/inputsource/common/streaming/listener.go [hunk 7]
```diff
@@ -202,24 +229,6 @@ func SplitFunc(framing FramingType, lineDelimiter []byte) (bufio.SplitFunc, erro
 	case FramingRFC6587:
 		return FactoryRFC6587Framing(lineDelimiter), nil
 	default:
-		return nil, fmt.Errorf("unknown SplitFunc for framing %d and line delimiter %s", framing, string(lineDelimiter))
+		return nil, fmt.Errorf("unknown SplitFunc for framing %d and line delimiter %q", framing, string(lineDelimiter))
```
Comment (efd6) [inline filebeat/inputsource/common/streaming/listener.go:232]: string conversion is not required here.

**Review by efd6 — APPROVED** [review]

**Pull request merged**